pkg/engine/planner/physical/optimizer.go

```diff
@@ -261,6 +269,144 @@ func (r *projectionPushdown) applyProjectionPushdown(node Node, projections []Co
 
 var _ rule = (*projectionPushdown)(nil)
 
+// parseKeysPushdown determines which keys need to be parsed based on downstream operations
+type parseKeysPushdown struct {
+	plan *Plan
+}
+
+// apply implements rule.
+func (r *parseKeysPushdown) apply(node Node) bool {
```

Comment:
I think the optimization pass here is going to break for logs queries, where they always need all keys. This may need similar logic for logs queries like what's done for the projection pushdown.

(But I think treating the parse key pushdown as a projection pushdown still makes sense, and then you could reuse the existing logic for handling logs queries there)